internal/venice-client-common/src/main/java/com/linkedin/venice/compression/ZstdWithDictCompressor.java

```diff
@@ -37,38 +45,100 @@ public byte[] compress(byte[] data) {
   }
 
   @Override
-  public ByteBuffer compress(ByteBuffer data) {
+  public ByteBuffer compress(ByteBuffer data, int startPositionOfOutput) throws IOException {
+    long maxDstSize = Zstd.compressBound(data.remaining());
+    if (maxDstSize > Integer.MAX_VALUE) {
+      throw new ZstdException(Zstd.errGeneric(), "Max output size is greater than Integer.MAX_VALUE");
+    }
+    int sizeOfOutput = (int) maxDstSize + startPositionOfOutput;
     if (data.isDirect()) {
       // TODO: It might be a decent refactor to add a pool of direct memory buffers so as to always leverage the this
       // interface and copy the results of the compression back into the passed in ByteBuffer. That would avoid
       // some of the data copy going on here.
-      return compressor.get().compress(data);
+      ByteBuffer output = ByteBuffer.allocateDirect(sizeOfOutput);
+      output.position(startPositionOfOutput);
+      data.mark();
+      int size = compressor.get().compress(output, data);
+      output.position(startPositionOfOutput);
+      output.limit(startPositionOfOutput + size);
+      data.reset();
+      return output;
     } else {
-      return ByteBuffer.wrap(compressor.get().compress(ByteUtils.extractByteArray(data)));
+      byte[] dst = new byte[sizeOfOutput];
+      int size = compressor.get()
+          .compressByteArray(
+              dst,
+              startPositionOfOutput,
+              (int) maxDstSize,
+              data.array(),
```

Comment:
Hmm, are you sure? This is a NIO BB, not a Netty BB. I know Netty has `CompositeByteBuf` but I'm not aware of such a thing in NIO. Do you know of a NIO BB subclass that has >1 array?

Reply:
Same as above, and there is no such implementation in NIO, but `ByteBuffer` is just an abstract class, which means people could have many other implementations.
If our assumption is that if the ByteBuffer is not direct, it must be HeapByteBuffer, which is backed by a single array, shall we do a type check to be safe?

Reply:
I suppose I could refactor this slightly so that it looks like:

```
if (data.isDirect()) {
  // same as above
} else if (data.hasArray()) {
  // same as above
} else {
  // throw
}
```

It's not clear to me what the semantics are for handling a non-direct and non-heap BB, so I would just throw defensively in that case. What do you think?

Reply:
Sounds good.
What I am mainly looking for is the safeguard and in case there are other implementation like `CompositeByteBuffer`, it is good to make sure the code doesn't fail siliently.

Reply:
I've added this in the latest commit.